### PR TITLE
populate tree with folder enumeration

### PR DESCRIPTION
**Pull request by ryanfkeepers**
adds basic parity for folder enumeration using the new delta tree struct.  Focuses only on handling new folder additions from delta enumeration.  PrevPath handling and item additions will come in later PRs.

PR bookend for:
* https://github.com/alcionai/corso/pull/4719

---

#### Does this PR need a docs update or release note?

- [x] :no_entry: No

#### Type of change

- [x] :sunflower: Feature

#### Issue(s)

* #4689

#### Test Plan

- [x] :zap: Unit test

### CODE REVIEW

**Review by ashmrtn — COMMENTED** [review]
Mostly nits, but would like to see a plan and/or update for the limits stuff prior to merging

File: src/internal/m365/collection/drive/collections_tree.go
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections_tree.go:278]: hum, that's a good question.  Can I hand wave and say, "I'll definitely figure something out because I'm going to have to have something figured out"?

File: src/internal/m365/collection/drive/collections_tree.go
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections_tree.go:270]: fair point.

File: src/internal/m365/collection/drive/collections_tree.go
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections_tree.go:378]: Also present in the limiter refactor changes.

File: src/internal/m365/collection/drive/collections_tree.go
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections_tree.go:499]: Good catch, yes, thank you.

File: src/internal/m365/collection/drive/collections_tree_test.go
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections_tree_test.go:483]: done

File: src/internal/m365/collection/drive/collections_tree_test.go
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections_tree_test.go:645]: Punting: I've got a big refactor to these constructors coming up later.  I don't address the drive IDs in that specifically, but it does remove a lot of the `parent(...)` chaff.  Happy to look at improving how we reference drives sometime after that, because I agree the current version isn't entirely obvious.

File: src/internal/m365/collection/drive/collections_tree_test.go
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections_tree_test.go:640]: sure, can do

File: src/internal/m365/collection/drive/collections_tree_test.go
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections_tree_test.go:771]: I've been noodling around where to put that documentation.  There's no clear owner for it, so I don't have one clear spot where I can explain the hidden expectations in a way that's easily discoverable for all maintainers.  If you have any ideas I'm open to them.

File: src/internal/m365/collection/drive/collections_tree.go
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections_tree.go:385]: fair; now returning the skipped value and reporting it in this func.

File: src/pkg/count/keys.go
Comment (ryanfkeepers) [reply src/pkg/count/keys.go:69]: removed tombstones, since it's not currently getting used.  I'll re-add it in the post-processing step.

In general, `TotalFooProcessed` is the count of "the number of times we saw this thing, strictly increasing even across resets and multiple deltas and etc; it's not a count of results, it's a count of the times we processed a Foo".

File: src/internal/m365/collection/drive/collections_tree_test.go
Comment (ashmrtn) [reply src/internal/m365/collection/drive/collections_tree_test.go:968]: this isn't so much about porting to other services so much as hiding details from tests and other non-tree code for the sake of maintainability. The more we access internal fields the more we'll need to update in the future if we ever rename/modify/refactor the tree itself

File: src/internal/m365/collection/drive/collections_tree_test.go
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections_tree_test.go:980]: Right, right.  The general idea I'm failing to follow is that higher layers only need to consider the behaviors that are unique to them.  Eg: this test should focus on single-page enumeration.  PopulateTree should focus on multi-page (and, later, multi-delta) enumeration.  I'm not doing a very good job at the separation right now.  I can go clean up the duplicated funcs before we merge this.

Comment (ashmrtn) [reply src/internal/m365/collection/drive/collections_tree_test.go:980]: That's one way to put it. Another way to view it is that these tests can be removed altogether because errors in this code will manifest as test failures for the higher-level API/functions that eventually call these functions

The tradeoff between the approaches is mostly fine-grained debugging vs. maintainability as it can sometimes be easier to maintain higher level tests against a stable API rather than fine-grained tests that may have churn due to implementation updates. API-level tests should also be portable across implementations (i.e. rewrite everything under the API) whereas the fine-grained tests won't be

Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections_tree_test.go:844]: okay, went through and removed some of the unnecessary test duplications across layers.  I'm going to keep the node checks at all levels, since I think it's better to have the assurance on the result state.  I'll revisit this reduction in future PRs as we go along as part of cascading the updates.

File: src/internal/m365/collection/drive/collections_tree_test.go
Comment (ashmrtn) [inline src/internal/m365/collection/drive/collections_tree_test.go:377]: nit: is this a stale comment that can be removed?

File: src/internal/m365/collection/drive/collections_tree_test.go
Comment (ashmrtn) [inline src/internal/m365/collection/drive/collections_tree_test.go:485]: nit: minor spelling fix
```suggestion
// More granular testing can be found in the lower level test functions below.
```

File: src/internal/m365/collection/drive/collections_tree_test.go
Comment (ashmrtn) [reply src/internal/m365/collection/drive/collections_tree_test.go:691]: hmmm maybe some documentation at the definition for the counter keys as to what "Processed" means (similar to what you added in another comment about those keys) with a note that some things may be "processed" multiple times due to how the external service returns data? At least that gives us a starting point for people who look at counter keys

As for the documentation for this particular test, I'm not really sure how to make it easier to grok. I'd like to avoid repeating comments for each test, but a comment on the test itself seems a bit too high-level

**Review by ashmrtn — APPROVED** [review]
just a few minor nits

File: src/internal/m365/collection/drive/collections_tree_test.go
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections_tree_test.go:485]: bBrgle

File: src/internal/m365/collection/drive/collections_tree_test.go
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections_tree_test.go:691]: done

File: src/internal/m365/collection/drive/collections_tree_test.go
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections_tree_test.go:377]: it'll come back later.  I have more assertions than necessary for the test atm, because it's not a fully implemented test.  Was easier to comment things out than to fix the assertions.

**Pull request merged**
